tests/project_helpers.py

```diff
@@ -22,9 +24,91 @@
 
 JsonData = Dict[str, Any]
 
+PACKAGES_TOML_PATH = Path(__file__).with_name("python_packages.toml")
+
 logger = logging.getLogger(__name__)
 
 
+@dataclass
+class TarballPackage:
+    """Encapsulate a Python tarball package.
+
+    This object fields are expected to be populated either from:
+    - A single toml file containing multiple tarballs info
+    - Multiple TOML files in REAL_PROJECTS_DIR
+    The tarballs are downloaded and cached by the methods below.
+    """
+
+    url: str
+    sha256: str
+    name: Optional[str] = None
+
+    @classmethod
+    def collect_from_toml(cls, path: Path) -> Iterator["TarballPackage"]:
+        """Parse information on all available tarball packages in a toml file."""
+        tarballs = parse_toml(path)
+        for info in tarballs.values():
+            yield cls(url=info["url"], sha256=info["sha256"])
```

Comment:
I've done exactly this in a previous iteration but the problem is that the presence of the name field will make the `tarball_name` function check if it's part of the downloaded tarball's name, and if not, it will change the downloaded tarball's name.
My reasoning is that unlike real world projects that might come from Github links or others, these packages tarballs come from PyPI-like repos and you need to keep the tarball's name as is for it to be correctly discovered. If you happen to mis-name your package in the toml file (see the `flict-core` example above), the tarball would be renamed and you'll not only lose the version information from the name but the package itself will be completely unrecogizable by pip.
Note that in the code below, I have added an additional test `if self.name` in the function to be able to still use it for real world projects while not having it change the name in the case of a native tarball where the name is None.
```python
def tarball_name(self) -> str:
        """The filename used for the tarball in the local cache."""
        # We cache tarballs using the filename part of the given URL.
        # However, tarballs produced from tags at GitHub typically only use the
        # version number in the filename. Prefix the project name in that case:
        filename = Path(urlparse(self.url).path).name
        if self.name and self.name not in filename:
            filename = f"{self.name}-{filename}"
        return filename
```
In another iteration, I also tried adding a parameter (e.g. `should_include_package_name`) to the `tarball_name` function that is set to default by False, but then didn't know how to use it differently in the case of a real world project.

Reply:
Ah I see now. That makes sense. I wonder if we could make this consideration a little more explicit in the code though, as omitting the `.name` on PyPI downloads as a way to prevent renaming is a little too opaque.

AFAICS, the `.name` is only ever used in `tarball_name()`, so we should be free to rename it to something more appropriate (along with an explanation in a comment). What about this?

```suggestion
    url: str
    sha256: str
    # If the following is given, we'll make sure the cached tarball has this
    # somewhere in its filename (helps with version-only tarballs from GitHub).
    filename_must_include: Optional[str] = None

    @classmethod
    def collect_from_toml(cls, path: Path) -> Iterator["TarballPackage"]:
        """Parse information on all available tarball packages in a toml file."""
        tarballs = parse_toml(path)
        for info in tarballs.values():
            yield cls(url=info["url"], sha256=info["sha256"])
```

Reply:
Applied the suggestion in [216cb26](https://github.com/tweag/FawltyDeps/pull/286/commits/216cb260660b5afc839ba46d961dcff07de543e7).